dual/provider/provider.go

```diff
@@ -0,0 +1,139 @@
+			provider.WithReprovideInterval(cfg.reprovideInterval[i]),
+			provider.WithMaxReprovideDelay(cfg.maxReprovideDelay[i]),
+			provider.WithConnectivityCheckOnlineInterval(cfg.connectivityCheckOnlineInterval[i]),
+			provider.WithConnectivityCheckOfflineInterval(cfg.connectivityCheckOfflineInterval[i]),
+			provider.WithMaxWorkers(cfg.maxWorkers[i]),
+			provider.WithDedicatedPeriodicWorkers(cfg.dedicatedPeriodicWorkers[i]),
+			provider.WithDedicatedBurstWorkers(cfg.dedicatedBurstWorkers[i]),
+			provider.WithMaxProvideConnsPerWorker(cfg.maxProvideConnsPerWorker[i]),
+		}
+		sweepingProviders[i], err = provider.New(dhtOpts...)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &SweepingProvider{
+		dht:      d,
+		LAN:      sweepingProviders[0],
+		WAN:      sweepingProviders[1],
+		keyStore: cfg.keyStore,
+	}, nil
+}
+
+// ProvideOnce only sends provider records for the given keys out to both
+// DHT swarms. It does NOT take the responsibility to reprovide these keys.
+func (s *SweepingProvider) ProvideOnce(keys ...mh.Multihash) {
+	go s.LAN.ProvideOnce(keys...)
+	go s.WAN.ProvideOnce(keys...)
+}
+
+// StartProviding provides the given keys to both DHT swarms unless they are
+// currently being reprovided. The keys will be periodically reprovided until
+// StopProviding is called for the same keys or user defined garbage collection
+// deletes the keys.
+func (s *SweepingProvider) StartProviding(force bool, keys ...mh.Multihash) {
+	ctx := context.Background()
+	newKeys, err := s.keyStore.Put(ctx, keys...)
+	if err != nil {
+		rLogger.Errorf("failed to store multihashes: %v", err)
+		return
+	}
+
+	s.LAN.AddToSchedule(newKeys...)
+	s.WAN.AddToSchedule(newKeys...)
+
+	if !force {
+		keys = newKeys
+	}
+
+	go s.ProvideOnce(keys...)
+}
+
+// StopProviding stops reproviding the given keys to both DHT swarms. The node
+// stops being referred as a provider when the provider records in both DHT
+// swarms expire.
+func (s *SweepingProvider) StopProviding(keys ...mh.Multihash) {
+	err := s.keyStore.Delete(context.Background(), keys...)
+	if err != nil {
+		rLogger.Errorf("failed to stop providing keys: %s", err)
+	}
+}
+
+// Clear clears the all the keys from the provide queues of both DHTs and returns the number
+// of keys that were cleared (sum of both queues).
+//
+// The keys are not deleted from the keystore, so they will continue to be
+// reprovided as scheduled.
+func (s *SweepingProvider) Clear() int {
+	return s.LAN.Clear() + s.WAN.Clear()
+}
+
+// RefreshSchedule scans the KeyStore for any keys that are not currently
+// scheduled for reproviding. If such keys are found, it schedules their
+// associated keyspace region to be reprovided for both DHT providers.
+//
+// This function doesn't remove prefixes that have no keys from the schedule.
+// This is done automatically during the reprovide operation if a region has no
+// keys.
+func (s *SweepingProvider) RefreshSchedule() {
+	s.LAN.RefreshSchedule()
+	s.WAN.RefreshSchedule()
```

Comment:
A random question that I have is whether these operations that are now sequential could happen in parallel, in a goroutine, and we wait for them to finish. Would that be faster/better?

Reply:
Updated to make operations on both DHT providers run in parallel.